backend/bitsurance/bitsurance.go

```diff
@@ -0,0 +1,156 @@
+	}
+	return strings.Split(string(responseBody), "\n"), nil
+}
+
+// findInsuredAccounts compares `accountIds` and `insuredIds` and returns a slice
+// with the matching elements.
+//
+// `accountIds` is an ordered slice of strings, containing the hashes of the
+// requested wallet accounts xpubs
+// `insuredIds` is another ordered slice of strings, containing the hashes of
+// all the insured xpubs retrieved from the Bitsurance server.
+//
+// The complexity of the scanning is linear.
+// For details on the hashing please refer to `BitsuranceGetId`.
+// TODO add tests.
+func findInsuredAccounts(accountIds, insuredIds []string) []string {
+	var matching []string
+
+	// Initialize indices for both lists
+	i, j := 0, 0
+
+	// Traverse both lists while there are elements in both
+	for i < len(accountIds) && j < len(insuredIds) {
+		// Compare the current elements in both lists
+		switch {
+		case accountIds[i] == insuredIds[j]:
+			// If the elements match, add it to the matching slice
+			matching = append(matching, accountIds[i])
+			i++
+			j++
+		case accountIds[i] < insuredIds[j]:
+			// If the element in accountIds is smaller, move to the next element
+			i++
+		default:
+			// If the element in insuredIds is smaller, move to the next element
+			j++
+		}
+	}
+
+	return matching
+}
+
+// BitsuranceGetId returns the BitsuranceId of a given account.
+// The id is computed hashing with sha256 the P2WPKH xpub of the account, concatenated with a fixed salt.
+// If a P2WPKH xpub can't be found in the account, empty string is returned.
+func BitsuranceGetId(account accounts.Interface) string {
+	for _, signingConf := range account.Info().SigningConfigurations {
+		bitcoinScriptType := signingConf.BitcoinSimple
+		if bitcoinScriptType != nil && bitcoinScriptType.ScriptType == signing.ScriptTypeP2WPKH {
+			h := sha256.New()
+			h.Write([]byte(bitcoinScriptType.KeyInfo.ExtendedPublicKey.String() + xpubSalt))
+			return hex.EncodeToString(h.Sum(nil))
+		}
+	}
+	return ""
+
+}
+
+// BitsuranceURL reutrns the url for the Bitsurance widget for a given locale.
+func BitsuranceURL(lang, bitsuranceId string) string {
+	return widgetURL + lang + "/?id=" + bitsuranceId
+}
+
+// BitsuranceAccountsLookup takes in input a slice of accounts. It fetches the list of insured accounts from the
+// Bitsurance server and returns a map with the given accounts' codes as keys and bool values which are `true` if
+// the given account is found on the list, `false` otherwise.
+func BitsuranceAccountsLookup(accounts []accounts.Interface, httpClient *http.Client) (map[types.Code]bool, error) {
+	insuredAccountsMap := make(map[types.Code]bool)
+	idCodeMap := make(map[string]types.Code)
+	var accountIds []string
+
+	bitsuranceIds, err := fetchBitsuranceIds(httpClient)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, account := range accounts {
+		code := account.Config().Config.Code
+		accountId := BitsuranceGetId(account)
+		idCodeMap[accountId] = code
```

Comment:
you probably want to skip this if accountId is empty, otherwise there could be clashes with multiple empty ones, and also potentially flag ETH accounts as insured if the bitsurance ID list contains a non-filtered empty line.